lib/ar_state_machine.rb

```diff
@@ -0,0 +1,272 @@
+require "ar_state_machine/version"
+require "active_record"
+
+module ARStateMachine
+  extend ActiveSupport::Concern
+
+  included do
+    cattr_accessor    :states
+
+    attr_accessor     :last_edited_by_id
+    attr_accessor     :skipped_transition
+
+    after_initialize  :state_machine_set_initial_state
+    before_update     :do_state_change_before_callbacks, 
+                      if: "state_changed? or (skipped_transition and skipped_transition.to_s == state.to_s)"
+    after_update      :do_state_change_do_after_callbacks, 
+                      if: "state_changed? or (skipped_transition and skipped_transition.to_s == state.to_s)"
+    before_update     :save_state_change, 
+                      if: "state_changed? or (skipped_transition and skipped_transition.to_s == state.to_s)"
+    validate          :state_machine_validation
+    validates         :state,
+                      presence: true
+
+    has_many          :state_changes, as: :source,
+                      dependent: :delete_all
+  end
+
+  private
+
+  def old_state
+    old_state = self.changed_attributes['state']
+
+    # we usually only want to create the state change if the state actually changes but
+    #   we also want to create it if it fails to transition from a state and stays there
+    #   example: a SubscriptionSeason in errored attempts to move to purchased, the order 
+    #   fails to build and it goes back to errored. We want to log the attempt.
+    #   we also want to do any callbacks
+    old_state ||= skipped_transition if skipped_transition.to_s == state.to_s
+    
+    old_state
+  end
+
+  def do_state_change_do_after_callbacks
+    self.class.run_after_transition_callbacks(self.state, self, old_state) if self.state_changed?
+  end
+
+  def save_state_change
+    self.state_changes.create({
+      previous_state: old_state,
+      next_state:     self.state,
+      created_by_id:  self.last_edited_by_id || User::SYSTEM_ID
+    })
+  end
+
+  def do_state_change_before_callbacks
+    rollback = self.class.run_before_transition_callbacks(self.state, self, old_state)
+    if self.skipped_transition and self.respond_to?("#{self.skipped_transition}_at=")
+      self.send("#{self.skipped_transition}_at=", Time.now)
```

Comment:
weird. Time.zone.now is an active support feature and should be available since you required active_record  http://api.rubyonrails.org/classes/ActiveSupport/TimeZone.html